### ISSUE TITLE
Regression: X509_sign, etc., no longer implicitly refresh the cached TBSCertificate

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#19388]
#18879 regressed `i2d_re_X509_REQ_tbs` by setting the modified bit after, rather than before, serializing the TBS structure, which was fixed in #19299. It looks like that wasn't the only such regression. It did the same to `X509_sign` and friends.

The changes in that PR to x_all.c should be reverted. The aim with setting the modified bit in `X509_sign` is not to mark that something changed, but to refresh the cached TBS structure before signing it. (OpenSSL's `X509` API returns too many mutable pointers to properly manage that dirty bit automatically, alas.)

(Marking the TBS structure dirty after adding a signature doesn't make sense because the signature isn't in the TBS structure.)

The practical implications here are that code which starts with an existing, parsed `X509` object as template (so the cache is filled), then modifies it and re-signs it, will break in OpenSSL 1.1.1r and 3.0.6. Likewise with the other crypto/x509 types.

**Comment by gibeom-gwon** [comment]
Sorry for my mistake. I thought setting the modified bit should be done after signing.
It looks like these commits need to be reverted.
- 9249a34b076df9a9d55ab74ab465d336980cae6a from master
- 9bc8fe33e15610538f48010f943c9a603b81283b from openssl-3.0
- 748df1874f0488ce0c86b6d2d083921abb34b1e3 from OpenSSL_1_1_1-stable

Should I prepare PRs for this?

**Comment by t8m** [comment]
Unfortunately this report did not surface yesterday.

Apparently we cannot release without at least one regression...

**Comment by davidben** [comment]
> Unfortunately this report did not surface yesterday.

Alas, I only noticed it while reviewing the diff between the releases. :-(

**Comment by slontis** [comment]
As @davidben has already mentioned, I think this may be showing us that even changing small lines without having tests is a really bad idea. Reviewers may need to be a bit more vigilant in this regard.

**Comment by slontis** [comment]
@davidben I am searching for access to this field.. I can see places where it sets it to 1 and one place where is uses it in X509_cmp(). But I dont see where it gets cleared. (Maybe my search is not working),,

**Comment by t8m** [comment]
Well, adding tests to the pr that changed these lines would not help because the contributor as well as both reviewers did not realize what was the real intent of the API. Of course writing tests when this functionality was added at the beginning would help but that simply did not happen. Given the huge API surface of OpenSSL it is impossible to catch this kind of regressions before someone actually tries to use the code.

And yeah, an argument could be - do not ever change code in stable release that does not already have thorough test coverage but that would make much bug fixing in stable releases impossible.

What might partially help is to always create release candidate releases in stable series and do the actual release only after a week or so of having the rc release in public. Of course if users won't bother reviewing and testing the rc releases, regressions would still slip through.

**Comment by slontis** [comment]
I am not quite understanding what is going on here... I wrote some test code...
````
static const unsigned char cert[] = {
    0x30, 0x82, 0x01, 0x86, 0x30, 0x82, 0x01, 0x2d, 0x02, 0x14, 0x75, 0xd6, 0x04, 0xd2, 0x80, 0x61,
    0xd3, 0x32, 0xbc, 0xae, 0x38, 0x58, 0xfe, 0x12, 0x42, 0x81, 0x7a, 0xdd, 0x0b, 0x99, 0x30, 0x0a,
    0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x04, 0x03, 0x02, 0x30, 0x45, 0x31, 0x0b, 0x30, 0x09,
    0x06, 0x03, 0x55, 0x04, 0x06, 0x13, 0x02, 0x41, 0x55, 0x31, 0x13, 0x30, 0x11, 0x06, 0x03, 0x55,
    0x04, 0x08, 0x0c, 0x0a, 0x53, 0x6f, 0x6d, 0x65, 0x2d, 0x53, 0x74, 0x61, 0x74, 0x65, 0x31, 0x21,
    0x30, 0x1f, 0x06, 0x03, 0x55, 0x04, 0x0a, 0x0c, 0x18, 0x49, 0x6e, 0x74, 0x65, 0x72, 0x6e, 0x65,
    0x74, 0x20, 0x57, 0x69, 0x64, 0x67, 0x69, 0x74, 0x73, 0x20, 0x50, 0x74, 0x79, 0x20, 0x4c, 0x74,
    0x64, 0x30, 0x20, 0x17, 0x0d, 0x32, 0x32, 0x31, 0x30, 0x31, 0x32, 0x30, 0x37, 0x32, 0x37, 0x35,
    0x35, 0x5a, 0x18, 0x0f, 0x32, 0x30, 0x35, 0x30, 0x30, 0x32, 0x32, 0x37, 0x30, 0x37, 0x32, 0x37,
    0x35, 0x35, 0x5a, 0x30, 0x45, 0x31, 0x0b, 0x30, 0x09, 0x06, 0x03, 0x55, 0x04, 0x06, 0x13, 0x02,
    0x41, 0x55, 0x31, 0x13, 0x30, 0x11, 0x06, 0x03, 0x55, 0x04, 0x08, 0x0c, 0x0a, 0x53, 0x6f, 0x6d,
    0x65, 0x2d, 0x53, 0x74, 0x61, 0x74, 0x65, 0x31, 0x21, 0x30, 0x1f, 0x06, 0x03, 0x55, 0x04, 0x0a,
    0x0c, 0x18, 0x49, 0x6e, 0x74, 0x65, 0x72, 0x6e, 0x65, 0x74, 0x20, 0x57, 0x69, 0x64, 0x67, 0x69,
    0x74, 0x73, 0x20, 0x50, 0x74, 0x79, 0x20, 0x4c, 0x74, 0x64, 0x30, 0x59, 0x30, 0x13, 0x06, 0x07,
    0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01,
    0x07, 0x03, 0x42, 0x00, 0x04, 0xbf, 0x82, 0xd9, 0xc9, 0x4b, 0x19, 0x43, 0x45, 0x6b, 0xd4, 0x50,
    0x64, 0x9b, 0xd5, 0x8d, 0x5a, 0xd9, 0xdc, 0xc9, 0x24, 0x23, 0x7a, 0x3b, 0x48, 0x23, 0xe2, 0x2a,
    0x24, 0xf2, 0x9c, 0x6f, 0x87, 0xd0, 0xc4, 0x0f, 0xcc, 0x7e, 0x7c, 0x8d, 0xfc, 0x08, 0x46, 0x37,
    0x85, 0x4f, 0x5b, 0x3a, 0x0b, 0x97, 0xd7, 0x57, 0x2a, 0x5a, 0x6b, 0x7a, 0x0b, 0xe4, 0xe8, 0x9c,
    0x4a, 0xbb, 0xbf, 0x09, 0x4d, 0x30, 0x0a, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x04, 0x03,
    0x02, 0x03, 0x47, 0x00, 0x30, 0x44, 0x02, 0x20, 0x5f, 0x45, 0x7f, 0xa4, 0x6a, 0x03, 0xfd, 0xe7,
    0xf3, 0x42, 0x43, 0x38, 0x5b, 0x81, 0x08, 0x1a, 0x47, 0x8e, 0x59, 0x3a, 0x28, 0x5b, 0x97, 0x67,
    0x47, 0x66, 0x2a, 0x16, 0xf5, 0xce, 0xf5, 0x92, 0x02, 0x20, 0x22, 0x0e, 0xab, 0x35, 0xdf, 0x49,
    0xb1, 0x86, 0xa3, 0x3b, 0x26, 0xda, 0x7e, 0x8b, 0x44, 0x45, 0xc6, 0x46, 0x14, 0x04, 0x22, 0x2b,
    0xe5, 0x2a, 0x62, 0x84, 0xc5, 0x94, 0xa0, 0x1b, 0xaa, 0xa9
};

static const unsigned char privkey[] = {
    0x30, 0x77, 0x02, 0x01, 0x01, 0x04, 0x20, 0x7d, 0x2b, 0xfe, 0x5c, 0xcb, 0xcb, 0x27, 0xd6, 0x28,
    0xfe, 0x98, 0x34, 0x84, 0x4a, 0x13, 0x6f, 0x70, 0xc4, 0x1a, 0x0b, 0xfc, 0xde, 0xb0, 0xb2, 0x32,
    0xb1, 0xdd, 0x4f, 0x0e, 0xbc, 0xdf, 0x89, 0xa0, 0x0a, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d,
    0x03, 0x01, 0x07, 0xa1, 0x44, 0x03, 0x42, 0x00, 0x04, 0xbf, 0x82, 0xd9, 0xc9, 0x4b, 0x19, 0x43,
    0x45, 0x6b, 0xd4, 0x50, 0x64, 0x9b, 0xd5, 0x8d, 0x5a, 0xd9, 0xdc, 0xc9, 0x24, 0x23, 0x7a, 0x3b,
    0x48, 0x23, 0xe2, 0x2a, 0x24, 0xf2, 0x9c, 0x6f, 0x87, 0xd0, 0xc4, 0x0f, 0xcc, 0x7e, 0x7c, 0x8d,
    0xfc, 0x08, 0x46, 0x37, 0x85, 0x4f, 0x5b, 0x3a, 0x0b, 0x97, 0xd7, 0x57, 0x2a, 0x5a, 0x6b, 0x7a,
    0x0b, 0xe4, 0xe8, 0x9c, 0x4a, 0xbb, 0xbf, 0x09, 0x4d
};

#include <openssl/x509.h>
#include "crypto/x509.h"

int main(void)
{
    int ret = 0;
    unsigned char enc[1024];
    X509 *x = NULL;
    const X509_ALGOR *sig_alg;
    const ASN1_BIT_STRING *sig;
    BIO *bio = BIO_new_fp(stdout, BIO_NOCLOSE | BIO_FP_TEXT);
    EVP_PKEY *pkey = NULL;
    EVP_MD *md = NULL;
    const unsigned char *keydata = privkey;
    const unsigned char *cdata = cert;

    md = EVP_MD_fetch(NULL, "SHA256", NULL);
    if (md == NULL)
        goto cleanup;

    pkey = d2i_PrivateKey(EVP_PKEY_EC, NULL, &keydata, sizeof(privkey));
    if (pkey == NULL)
        goto cleanup;
    x = d2i_X509(NULL, &cdata, sizeof(cert));
    if (x == NULL)
        goto cleanup;
    memcpy(enc, x->cert_info.enc.enc, x->cert_info.enc.len);

    X509_sign(x, pkey, md);

    if (memcmp(x->cert_info.enc.enc, enc, x->cert_info.enc.len) != 0)
        goto cleanup;

    time_t now = time(NULL);
    /* Pick a day earlier and later, relative to any system clock. */
    ASN1_TIME *asn1_after = ASN1_TIME_adj(NULL, now, 1000, 0);
    X509_set1_notAfter(x, asn1_after);
    X509_sign(x, pkey, md);
    if (memcmp(x->cert_info.enc.enc, enc, x->cert_info.enc.len) == 0)
        goto cleanup;
````
I was expecting the code to behave differently when I swapped between the 2 implementations of X509_sign() (With the enc.modified = 1 set in 2 places).. I would of thought that it would recache when you changed a value, but I dont think it does,,

**Comment by slontis** [comment]
Both case seem to fail on the last memcmp

**Comment by davidben** [comment]
@slontis The `modified` bit there is a little interesting. Apologies, I probably confused this because I described it as "refreshing" the cache, when it's actually "ignoring the cache". The i2d functions do not refresh the cached encoding. (If they did, it'd be yet another place where i2d_Foo is not thread-safe!)

Rather, the logic is:
* If there is a valid, cached encoding, replay it
* Otherwise, serialize the structure de novo, don't cache it
https://github.com/openssl/openssl/blob/master/crypto/asn1/tasn_enc.c#L152

This works out because if we're serializing de novo and signing, we can assume that serializing it again will still produce the same result. Whereas, if we got the `X509` out of `d2i_X509`, because OpenSSL incorrectly allows BER, re-encoding the TBSCertificate may not produce the same bytes. (Actually RFC 5280 says you're supposed to [re-encode as DER](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.1.3) to check the signature, but nobody does this.)

That means writing a test by inspecting `enc` won't do anything useful. Especially as this internal representation is super goofy, I think it's better to write these tests based on observable effects. Here's a series of tests I wrote after #18879 and now this. At least in BoringSSL, they're sensitive to both the previous regression and this one.
https://boringssl.googlesource.com/boringssl/+/da96eeb958ee8442f408a5e8fa78cbfb86cc3281/crypto/x509/x509_test.cc#1946

(Those tests also use the confusing "refresh" terminology. Apparently I keep forgetting that little detail. I've uploaded a change to switch that to "drop".)
